lib/node_runner.go

```diff
@@ -186,93 +199,267 @@ func (nr *NodeRunner) SetHandleBallotCheckerFuncs(
-				LocalNode:              nr.localNode,
-				NetworkID:              nr.networkID,
-				Message:                message,
-				VotingHole:             VotingNOTYET,
-			}
-			if err = sebakcommon.RunChecker(checker, nr.handleBallotCheckerDeferFunc); err != nil {
-				if _, ok := err.(sebakcommon.CheckerErrorStop); !ok {
-					nr.log.Error("failed to handle ballot", "error", err)
-				}
-			}
-			nr.closeConsensus(checker)
-		default:
-			nr.log.Error("got unknown", "message", message.Head(50))
+	if err = sebakcommon.RunChecker(checker, nr.handleMessageFromClientCheckerDeferFunc); err != nil {
+		if _, ok := err.(sebakcommon.CheckerErrorStop); ok {
+			return
+		}
+		nr.log.Error("failed to handle message from client", "error", err)
+		return
+	}
+
+	return
+}
+
+func (nr *NodeRunner) handleBallotMessage(message sebaknetwork.Message) (err error) {
+	nr.log.Debug("got ballot", "message", message.Head(50))
+
+	checker := &NodeRunnerHandleBallotChecker{
+		GenesisBlockCheckpoint: sebakcommon.MakeGenesisCheckpoint(nr.networkID),
+		DefaultChecker:         sebakcommon.DefaultChecker{nr.handleBallotCheckerFuncs},
+		NodeRunner:             nr,
+		LocalNode:              nr.localNode,
+		NetworkID:              nr.networkID,
+		Message:                message,
+		VotingHole:             VotingNOTYET,
+	}
+	err = sebakcommon.RunChecker(checker, nr.handleBallotCheckerDeferFunc)
+	if err != nil {
+		if _, ok := err.(sebakcommon.CheckerErrorStop); ok {
+			nr.log.Debug("stop handling ballot", "reason", err)
+		} else {
+			nr.log.Debug("failed to handle ballot", "error", err)
 		}
 	}
+
+	return
 }
 
-func (nr *NodeRunner) closeConsensus(c sebakcommon.Checker) (err error) {
-	checker := c.(*NodeRunnerHandleBallotChecker)
+func (nr *NodeRunner) InitRound() {
+	// get latest blocks
+	var err error
+	var latestBlock Block
+	if latestBlock, err = GetLatestBlock(nr.storage); err != nil {
+		panic(err)
+	}
 
-	if checker.VotingStateStaging.IsEmpty() {
-		return
+	nr.consensus.SetLatestConsensusedBlock(latestBlock)
+	nr.consensus.SetLatestRound(Round{})
+
+	ticker := time.NewTicker(time.Millisecond * 5)
+	for _ = range ticker.C {
+		var notFound bool
+		connected := nr.connectionManager.AllConnected()
+		for address, _ := range nr.localNode.GetValidators() {
+			if _, found := sebakcommon.InStringArray(connected, address); !found {
+				notFound = true
+				break
+			}
+		}
+		if !notFound {
+			ticker.Stop()
+			break
+		}
 	}
-	if !checker.VotingStateStaging.IsClosed() {
+
+	nr.log.Debug("all validators are checked for connectivity")
+
+	go nr.startRound()
+}
+
+func (nr *NodeRunner) startRound() {
+	// check whether current running rounds exist
+	if len(nr.consensus.RunningRounds) > 0 {
 		return
 	}
 
-	if err = nr.Consensus().CloseConsensus(checker.Ballot); err != nil {
-		nr.Log().Error("new failed to close consensus", "error", err)
+	nr.StartNewRound(0)
+}
+
+func (nr *NodeRunner) CalculateProposer(blockHeight uint64, roundNumber uint64) string {
```

Comment:
Unfortunately, I found this function is not fair for candidates.
I make a testcase like below.
```
func TestNodeRunnerCalculateProposer(t *testing.T) {
	defer sebaknetwork.CleanUpMemoryNetwork()

	numberOfNodes := 3
	nodeRunners := createTestNodeRunnerWithReady(numberOfNodes)

	nr0 := nodeRunners[0]

	proposer00 := nr0.CalculateProposer(0, 0)
	proposer01 := nr0.CalculateProposer(0, 1)
	proposer02 := nr0.CalculateProposer(0, 2)
	proposer03 := nr0.CalculateProposer(0, 3)
	proposer04 := nr0.CalculateProposer(0, 4)
	proposer05 := nr0.CalculateProposer(0, 5)

	if proposer00 == proposer01 {
		t.Error("failed to calculate correct proposers. proposer00", proposer00, "proposer01", proposer01)
	}
	if proposer00 == proposer02 {
		t.Error("failed to calculate correct proposers. proposer00", proposer00, "proposer02", proposer02)
	}
	if proposer01 == proposer02 {
		t.Error("failed to calculate correct proposers. proposer01", proposer01, "proposer02", proposer02)
	}
	if proposer00 != proposer03 {
		t.Error("failed to calculate correct proposers. proposer00", proposer00, "proposer03", proposer03)
	}
	if proposer01 != proposer04 {
		t.Error("failed to calculate correct proposers. proposer01", proposer01, "proposer04", proposer04)
	}
	if proposer02 != proposer05 {
		t.Error("failed to calculate correct proposers. proposer02", proposer02, "proposer05", proposer05)
	}

	proposer10 := nr0.CalculateProposer(1, 0)
	proposer11 := nr0.CalculateProposer(1, 1)
	proposer12 := nr0.CalculateProposer(1, 2)
	proposer13 := nr0.CalculateProposer(1, 3)
	proposer14 := nr0.CalculateProposer(1, 4)
	proposer15 := nr0.CalculateProposer(1, 5)

	if proposer10 == proposer11 {
		t.Error("failed to calculate correct proposers. proposer10", proposer10, "proposer11", proposer11)
	}
	if proposer10 == proposer12 {
		t.Error("failed to calculate correct proposers. proposer10", proposer10, "proposer12", proposer12)
	}
	if proposer11 == proposer12 {
		t.Error("failed to calculate correct proposers. proposer11", proposer11, "proposer12", proposer12)
	}
	if proposer10 != proposer13 {
		t.Error("failed to calculate correct proposers. proposer10", proposer10, "proposer13", proposer13)
	}
	if proposer11 != proposer14 {
		t.Error("failed to calculate correct proposers. proposer11", proposer11, "proposer14", proposer14)
	}
	if proposer12 != proposer15 {
		t.Error("failed to calculate correct proposers. proposer12", proposer12, "proposer15", proposer15)
	}

	for _, nr := range nodeRunners {
		nr.Stop()
	}
}
```
And I got messages like below.
```
--- FAIL: TestNodeRunnerCalculateProposer (3.32s)
	node_runner_test.go:180: failed to calculate correct proposers. proposer01 GAJKIJVMZKWQWAH74YOHTRREZT6CO6F2BVK6ABWXMWSM5SRLYNFIAO5T proposer04 GDSVSVTQF3LENGRR2VC2HWWFVVPD5HG2ITMKGFBOE2OLKRYZCL6OJ3CC
	node_runner_test.go:197: failed to calculate correct proposers. proposer10 GCCMSRGVCN5GZAP7YDRRRTYGGNDEA2AJKAFJOKTFI2PXQOE2WX3QSXXT proposer12 GCCMSRGVCN5GZAP7YDRRRTYGGNDEA2AJKAFJOKTFI2PXQOE2WX3QSXXT
	node_runner_test.go:203: failed to calculate correct proposers. proposer10 GCCMSRGVCN5GZAP7YDRRRTYGGNDEA2AJKAFJOKTFI2PXQOE2WX3QSXXT proposer13 GDSVSVTQF3LENGRR2VC2HWWFVVPD5HG2ITMKGFBOE2OLKRYZCL6OJ3CC
	node_runner_test.go:206: failed to calculate correct proposers. proposer11 GAJKIJVMZKWQWAH74YOHTRREZT6CO6F2BVK6ABWXMWSM5SRLYNFIAO5T proposer14 GDSVSVTQF3LENGRR2VC2HWWFVVPD5HG2ITMKGFBOE2OLKRYZCL6OJ3CC
	node_runner_test.go:209: failed to calculate correct proposers. proposer12 GCCMSRGVCN5GZAP7YDRRRTYGGNDEA2AJKAFJOKTFI2PXQOE2WX3QSXXT proposer15 GDSVSVTQF3LENGRR2VC2HWWFVVPD5HG2ITMKGFBOE2OLKRYZCL6OJ3CC
```
According to my analysis, proposers at block 1 and round 0, 1, 2 are same. Like this, this function seems to be more fairness.